File: sample-conference-videochat/sample-conference-videochat/Components/Calls/StreamInitiatorViewController/StreamInitiatorViewController.m
//
//  StreamInitiatorViewController.m
//  sample-conference-videochat
//
//  Created by Injoit on 24.03.2021.
//  Copyright © 2021 Quickblox. All rights reserved.
//

#import "StreamInitiatorViewController.h"
#import "StreamTitleView.h"

static const NSTimeInterval kRefreshTimeInterval = 2.0f;

@interface StreamInitiatorViewController ()
@property (strong, nonatomic) StreamTitleView *streamTitleView;
@property (strong, nonatomic) NSTimer *refreshLisnersTimer;
@property (assign, nonatomic) NSUInteger listenersCount;
@property (nonatomic, strong) id observerWillResignActive;
@property (nonatomic, strong) id observerWillActive;
@property (strong, nonatomic) NSNumber *currentConferenceUserID;
@end

@implementation StreamInitiatorViewController

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    
    [self setupNavigationBarWillAppear:NO];
    [self invalidateHideToolbarTimer];
    
    NSNotificationCenter *defaultCenter = [NSNotificationCenter defaultCenter];
    if (self.observerWillResignActive) {
        [defaultCenter removeObserver:(self.observerWillResignActive)];
    }
    if (self.observerWillActive) {
        [defaultCenter removeObserver:(self.observerWillActive)];
    }
}

- (void)didSetMuteVideo:(BOOL)muteVideo {
    self.session.localMediaStream.videoTrack.enabled = !muteVideo;
    [self.participants participantWithId:self.participants.localId].isCameraEnabled = !muteVideo;
    self.swapCamera.userInteractionEnabled = self.session.localMediaStream.videoTrack.enabled;
}

- (void)setupSession {
    if (!QBSession.currentSession.currentUser) {
        return;
    }
    self.currentConferenceUserID = @(QBSession.currentSession.currentUserID);
    QBUUser *currentUser = QBSession.currentSession.currentUser;
    
    // creating session
    NSString *conferenceID = self.conferenceSettings.conferenceInfo.conferenceID;
    self.session = [[QBRTCConferenceClient instance] createSessionWithChatDialogID:conferenceID conferenceType:QBRTCConferenceTypeVideo];
    if (!self.session) {
        return;
    }
    [self.participants addParticipantWithId:@(currentUser.ID) fullName:currentUser.name];
    
    NSNotificationCenter *center = NSNotificationCenter.defaultCenter;
    __weak __typeof(self)weakSelf = self;
    self.observerWillActive = [center addObserverForName:UIApplicationDidBecomeActiveNotification
                                                  object:nil
                                                   queue:NSOperationQueue.mainQueue
                                              usingBlock:^(NSNotification * _Nonnull note) {
        if (!weakSelf.cameraCapture) {return;}
        if ([weakSelf.participants participantWithId:weakSelf.participants.localId].isCameraEnabled != YES) {return;}
        
        weakSelf.session.localMediaStream.videoTrack.enabled = YES;
    }];
    
    self.observerWillResignActive = [center addObserverForName:UIApplicationWillResignActiveNotification
                                                  object:nil
                                                   queue:NSOperationQueue.mainQueue
                                              usingBlock:^(NSNotification * _Nonnull note) {
        if (!weakSelf.cameraCapture) {return;}
        weakSelf.session.localMediaStream.videoTrack.enabled = NO;
    }];
}

- (void)configureNavigationBarItems {
    self.streamTitleView = [[StreamTitleView alloc] init];
    self.navigationItem.titleView = self.streamTitleView;
    [self.streamTitleView setupStreamTitleViewOnLive:YES];
    
    self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"menu_chat"] style:UIBarButtonItemStylePlain target:self action:@selector(didTapChat:)];
    self.navigationItem.leftBarButtonItem.tintColor = UIColor.whiteColor;
    self.navigationItem.leftBarButtonItem.enabled = NO;
    
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"0 members" style:UIBarButtonItemStylePlain target:self action:nil];
    self.navigationItem.rightBarButtonItem.tintColor = UIColor.whiteColor;
    self.navigationItem.rightBarButtonItem.enabled = NO;
}

- (void)setupAudioVideoEnabledCell:(ConferenceUserCell *)cell forUserID:(NSNumber *)userID {
    if (userID == self.currentConferenceUserID) {
        self.session.localMediaStream.videoTrack.enabled = !self.muteVideo;
    }
    cell.isMuted = !self.muteAudio;
}

- (void)updateWithCreatedNewSession:(QBRTCConferenceSession *)session {
    self.navigationItem.rightBarButtonItem.enabled = YES;
    self.navigationItem.leftBarButtonItem.enabled = YES;
    
    QBRTCAudioSession *audioSession = [QBRTCAudioSession instance];
    
    QBRTCAudioSessionConfiguration *configuration = [[QBRTCAudioSessionConfiguration alloc] init];
    configuration.categoryOptions |= AVAudioSessionCategoryOptionDuckOthers;
    
    // adding blutetooth support
    configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowBluetooth;
    configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowBluetoothA2DP;
    
    // adding airplay support
    configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowAirPlay;
    
    configuration.mode = AVAudioSessionModeVideoChat;
    
    [audioSession setConfiguration:configuration active:YES];
    
    session.localMediaStream.audioTrack.enabled = YES;
    session.localMediaStream.videoTrack.videoCapture = self.cameraCapture;
    session.localMediaStream.videoTrack.enabled = NO;
    [session joinAsPublisher];
    
    [self createLisnersTimer];
}

- (void)createLisnersTimer {
    [self invalidateHideToolbarTimer];
    self.refreshLisnersTimer =[NSTimer scheduledTimerWithTimeInterval:kRefreshTimeInterval
                                                               target:self
                                                             selector:@selector(refreshCallTime)
                                                             userInfo:nil
                                                              repeats:YES];
    
}

- (void)invalidateLisnersTimer {
    if (self.refreshLisnersTimer != nil) {
        [self.refreshLisnersTimer invalidate];
        self.refreshLisnersTimer = nil;
    }
}

//MARK: - Internal Methods
- (void)refreshCallTime {
    [self updateNumberOfLisners];
}

- (void)updateNumberOfLisners {
    __weak __typeof(self)weakSelf = self;
    [self.session listOnlineParticipantsWithCompletionBlock:^(NSArray<NSNumber *> * _Nonnull publishers, NSArray<NSNumber *> * _Nonnull listeners) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if (strongSelf.listenersCount != listeners.count) {
            strongSelf.listenersCount = listeners.count;
            
            NSString *members = listeners.count == 1 ? @"member" : @"members";
            NSString *membersNumber = [NSString stringWithFormat:@"%@ %@", @(listeners.count), members];
            dispatch_async(dispatch_get_main_queue(), ^{
                strongSelf.navigationItem.rightBarButtonItem.title = membersNumber;
            });
        }
    }];
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Calls/StreamInitiatorViewController/StreamInitiatorViewController.swift
//
//  StreamInitiatorViewController.swift
//  sample-conference-videochat-swift
//
//  Created by Injoit on 16.08.2020.
//  Copyright © 2020 QuickBlox. All rights reserved.
//

import UIKit
import Quickblox
import QuickbloxWebRTC

struct StreamInitiatorViewControllerConstants {
    static let refreshTimeInterval: TimeInterval = 2.0
}

class StreamInitiatorViewController: BaseCallViewController {
    lazy private var streamTitleView: StreamTitleView = {
        let streamTitleView = StreamTitleView()
        return streamTitleView
    }()
    private var lisnersTimer: Timer?
    private var listenersCount = 0

    // MARK: Overrides
    override func configureNavigationBarItems() {
        navigationItem.titleView = streamTitleView
            streamTitleView.setupStreamTitleViewOnLive(true)
            
        navigationItem.rightBarButtonItem = UIBarButtonItem(title: "0 members",
                                                            style: .plain,
                                                            target: self,
                                                            action: nil)
        
        navigationItem.leftBarButtonItem = UIBarButtonItem(image: UIImage(named: "menu_chat"),
                                                           style: .plain,
                                                           target: self,
                                                           action: #selector(didTapChat(_:)))
        navigationItem.leftBarButtonItem?.tintColor = .white
        navigationItem.leftBarButtonItem?.isEnabled = false
        navigationItem.rightBarButtonItem?.tintColor = .white
        navigationItem.rightBarButtonItem?.isEnabled = false
    }
    
    override func setupAudioVideoEnabledCell(_ cell: ConferenceUserCell, forUserID userID: UInt) {
        cell.videoEnabled = true
        cell.unMute = !muteAudio
    }
    
    override func updateUIWithCreatedNewSession(_ session: QBRTCConferenceSession) {
        let audioSession = QBRTCAudioSession.instance()
        audioSession.initialize { configuration in
            // adding blutetooth support
            
            configuration.categoryOptions = .allowBluetoothA2DP
            configuration.categoryOptions = .allowBluetooth
            
            // adding airplay support
            configuration.categoryOptions = .allowAirPlay
            
            configuration.mode = AVAudioSession.Mode.videoChat.rawValue
        }
        
        session.localMediaStream.audioTrack.isEnabled = true
        session.localMediaStream.videoTrack.isEnabled = true
        session.localMediaStream.videoTrack.videoCapture = cameraCapture
        session.joinAsPublisher()
        createLisnersTimer()
    }
    
    override func showControls(_ isShow: Bool) {
        isShow == true ? createLisnersTimer() : invalidateLisnersTimer()
        setupControls(isShow)
    }
    
    func createLisnersTimer() {
        lisnersTimer = Timer.scheduledTimer(timeInterval: StreamInitiatorViewControllerConstants.refreshTimeInterval,
                                         target: self,
                                         selector: #selector(refreshCallTime(_:)),
                                         userInfo: nil,
                                         repeats: true)
    }
    
    func invalidateLisnersTimer() {
        if let callTimer = lisnersTimer {
            callTimer.invalidate()
            self.lisnersTimer = nil
        }
    }
    
    override func removeUserFromCollection(_ userID: NSNumber) {
        if let index = users.index(where: { $0.userID == userID.uintValue }) {
            users.remove(at: index)
        }
    }
    
    //MARK: - Internal Methods
    @objc private func refreshCallTime(_ sender: Timer?) {
        updateNumberOfLisners()
    }
    
    private func updateNumberOfLisners() {
        session?.listOnlineParticipants(completionBlock: { [weak self] publishers, listeners in
            guard let self = self, self.listenersCount != listeners.count else {return}
            self.listenersCount = listeners.count
            let members = listeners.count == 1 ? "member" : "members"
            DispatchQueue.main.async {
                self.navigationItem.rightBarButtonItem?.title = "\(self.listenersCount) " + members
            }
        })
    }
}
